core/src/services/icloud/backend.rs

```diff
@@ -0,0 +1,261 @@
+            None => Err(
+                Error::new(ErrorKind::ConfigInvalid, "ds_web_auth_token is empty")
+                    .with_operation("Builder::build")
+                    .with_context("service", Scheme::Icloud),
+            ),
+        }?;
+
+        let client = if let Some(client) = self.http_client.take() {
+            client
+        } else {
+            HttpClient::new().map_err(|err| {
+                err.with_operation("Builder::build")
+                    .with_context("service", Scheme::Icloud)
+            })?
+        };
+
+        let trust_token = match &self.trust_token {
+            Some(trust_token) => Ok(trust_token.clone()),
+            None => Err(Error::new(ErrorKind::ConfigInvalid, "trust_token is empty")
+                .with_operation("Builder::build")
+                .with_context("service", Scheme::Icloud)),
+        }?;
+        debug!("iCloud backend use trust_token {}", &trust_token);
+
+        let session_data = SessionData::new();
+
+        Ok(iCloudBackend {
+            client: Arc::new(Client {
+                session: Arc::new(Mutex::new(Session {
+                    data: session_data,
+                    client,
+                    apple_id,
+                    password,
+                    trust_token: Some(trust_token),
+                    ds_web_auth_token: Some(ds_web_auth_token),
+                })),
+                root,
+                path_cache: Arc::new(Default::default()),
+            }),
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct iCloudBackend {
+    client: Arc<Client>,
+}
+
+#[async_trait]
+impl Accessor for iCloudBackend {
+    type Reader = IncomingAsyncBody;
+    type BlockingReader = ();
+    type Writer = ();
+    type BlockingWriter = ();
+    type Lister = ();
+    type BlockingLister = ();
+
+    fn info(&self) -> AccessorInfo {
+        let mut ma = AccessorInfo::default();
+        ma.set_scheme(Scheme::Icloud)
+            .set_root(&self.client.root)
+            .set_native_capability(Capability {
+                stat: true,
+                read: true,
+                ..Default::default()
+            });
+        ma
+    }
+
+    async fn stat(&self, path: &str, _: OpStat) -> Result<RpStat> {
+        // iCloud get the filename by id, instead obtain the metadata by filename
+        if path == "/" {
+            return Ok(RpStat::new(Metadata::new(EntryMode::DIR)));
+        }
+
+        let node = self.client.stat(path).await?;
+
+        let mut meta = Metadata::new(match node.type_field.as_str() {
+            "FOLDER" => EntryMode::DIR,
+            _ => EntryMode::FILE,
+        });
+
+        if meta.mode() == EntryMode::DIR || path.ends_with('/') {
+            return Ok(RpStat::new(Metadata::new(EntryMode::DIR)));
+        }
+
+        meta = meta.with_content_length(node.size);
+
+        let last_modified = parse_datetime_from_rfc3339(&node.date_modified).map_err(|e| {
```

Comment:
All functions provided by `opendal::raw` handle errors correctly, so you can simply use `?` here instead.